runtime/src/comm/ofi/comm-ofi.c

```diff
@@ -5317,35 +5374,32 @@ void amPutDone(c_nodeid_t node, amDone_t* pAmDone) {
   }
 
   //
-  // Send the 'done' indicator.  If necessary, first make sure the
-  // memory effects of all communication ops done by this AM are
-  // visible.  We and the AM's initiator are the same Chapel task,
-  // for purposes of MCM conformance, and the memory effects of the
-  // AM must be visible to that initiator by the time it sees the
-  // 'done' indicator.  In the delivery-complete mode everything is
-  // already visible.  In the message-order mode we don't have to
-  // force PUT visibility because this 'done' is also a PUT and we
-  // use write-after-write message ordering, but we do have to force
-  // AMO visibility.  In message-order-fence mode we have to force
-  // both PUT and AMO visibility.
+  // Send the 'done' indicator.  If necessary, first make sure the memory
+  // effects of all communication ops to all nodes done by this AM are
+  // visible.  We and the AM's initiator are the same Chapel task, for
+  // purposes of MCM conformance, and the memory effects of the AM must be
+  // visible to that initiator by the time it sees the'done' indicator.  In
+  // the delivery-complete mode everything is already visible.  In the
+  // message-order and message-order-fence modes we have to force PUT and AMO
+  // visibility.
   //
   if (mcmMode != mcmm_dlvrCmplt) {
-    forceMemFxVisAllNodes(mcmMode == mcmm_msgOrdFence /*checkPuts*/,
-                          true /*checkAmos*/,
+    forceMemFxVisAllNodes(true /*checkPuts*/, true /*checkAmos*/,
```

Comment:
Is there the potential for perf issues here? Since are always forcing puts?

Reply:
Possibly in `message-order` MCM mode, which we don't really use, and it must be done to ensure MCM conformance. The MCM requires that all PUTs performed by an active message (AM) are visible when the active message completes. The function `amPutDone` does a PUT of a `done` variable to indicate that the AM is complete. Message ordering ensures that any pending PUTs to the same target locale will be visible before the PUT to `done` is visible; the problem is with PUTs to other locales.  For all MCM modes except for `delivery-complete` this requires an action to force visibility, which is what `forceMemFxVisAllNodes` does. Previously, it only checked PUTs when in `message-order-fence` mode, but it should do so for all MCM modes that are not `delivery-complete`.